shotover-proxy/src/transforms/redis/sink_cluster.rs

```diff
@@ -572,22 +587,51 @@ impl RoutingInfo {
+            // But I assume in such an environment the nodes would be configured to not allow running KEYS.
+            // So each redis node would return an error and then we would return a single error to the client which would be fine.
             b"KEYS" => RoutingInfo::AllMasters(ResponseJoin::ArrayJoin),
+            // The LASTSAVE command is needed to confirm that a previous BGSAVE command has succeed.
+            // In order to maintain this use case we query every node and return the oldest save time.
+            // This way the return value wont change until every node has completed their BGSAVE.
+            b"LASTSAVE" => RoutingInfo::AllNodes(ResponseJoin::IntegerMin),
+            // When a command that forces writing to disk occurs we want it to occur on every node.
+            // Replica nodes receive updates from their master nodes and we want those to be written to disk too.
+            b"BGSAVE" | b"SAVE" | b"BGREWRITEAOF" | b"ACL" => {
+                RoutingInfo::AllNodes(ResponseJoin::First)
+            }
             b"SCRIPT" => match args.get(1) {
                 Some(RedisFrame::BulkString(a)) if a.to_ascii_uppercase() == b"KILL" => {
                     RoutingInfo::Unsupported
                 }
                 _ => RoutingInfo::AllMasters(ResponseJoin::First),
             },
-            b"SCAN" | b"SHUTDOWN" | b"SLAVEOF" | b"REPLICAOF" | b"MOVE" | b"BITOP" => {
-                RoutingInfo::Unsupported
-            }
+            // * We cant reasonably support CLIENT SETNAME/GETNAME in shotover
+            //     - Connections are pooled so we cant forward it to the redis node
+            //     - We cant just implement the functionality shotover side because the connection names are supposed to be viewable from CLIENT LIST
+            // * However we dont want to just fail them either as clients like jedis would break so:
+            //     - We just pretend to accept CLIENT SETNAME returning an "OK" but without hitting any nodes
+            //     - We just pretend to handle CLIENT GETNAME always returning nil without hitting any nodes
+            b"CLIENT" => match args.get(1) {
+                Some(RedisFrame::BulkString(sub_command)) => {
+                    let sub_command = sub_command.to_ascii_uppercase();
+                    match sub_command.as_slice() {
+                        b"SETNAME" => RoutingInfo::ShortCircuitOk,
+                        b"GETNAME" => RoutingInfo::ShortCircuitNil,
+                        _ => RoutingInfo::Unsupported,
+                    }
+                }
+                _ => RoutingInfo::Random,
+            },
+            // These commands can not reasonably be supported by shotover, so we just return an error to the client when they are used
+            b"SCAN" | b"SHUTDOWN" | b"SLAVEOF" | b"REPLICAOF" | b"MOVE" | b"BITOP" | b"CONFIG"
+            | b"SLOWLOG" | b"INFO" | b"TIME" => RoutingInfo::Unsupported,
```

Comment:
Does redis-cli still work if we entirely forbid `CONFIG`? :/ I was planning to test it myself, but macOS + M1 has broken the (lack of) bridge networking workaround I was using, and I haven't had time to setup a Linux VM.

I thought of this because I remember it running CONFIG commands, but at a glance, I'm not sure whether it's required to succeed: https://github.com/redis/redis/blob/87131a5fa612c0d2b688fc65c722ebc95ae8f42a/src/redis-cli.c#L7043

Reply:
Should we have some basic client driver integration tests, to prevent regressions like #606?

(e.g. redis-cli + Lettuce + Jedis)

New ticket of course. :)

Reply:
this works:
![image](https://user-images.githubusercontent.com/5120858/166403571-b284ebb5-0bf9-490e-bb7f-40d87607683b.png)

this does not:
![image](https://user-images.githubusercontent.com/5120858/166403455-7f4618f2-5212-425e-88c5-c62377cb96ec.png)

Do we need to support --cluster though?
I dont think it makes sense when shotover is pretending to be a single non clustered redis instance.

Reply:
> Should we have some basic client driver integration tests, to prevent regressions like https://github.com/shotover/shotover-proxy/issues/606?

Yeah this would be really valuable if we can find a way to nicely pull in these non rust dependencies.

Reply:
> Do we need to support --cluster though?
> I dont think it makes sense when shotover is pretending to be a single non clustered redis instance.

For our managed service, cluster hiding is a bonus feature that isn't actually needed at the moment. The primary use case is Shotover mirroring for Redis clusters, so it's always `redis-cli` with `-c` in our docs/examples.

> find a way to nicely pull in these non rust dependencies

One more self-contained docker compose? :)

Reply:
> For our managed service, cluster hiding is a bonus feature that isn't actually needed at the moment. The primary use case is Shotover mirroring for Redis clusters, so it's always redis-cli with -c in our docs/examples.

Oh huh, I'll have to go over things again then.
When ben linked https://github.com/RedisLabs/redis-cluster-proxy/blob/unstable/COMMANDS.md I must have taken it too literally.

Reply:
I have confirmed this does work on main branch, so I'll investigate whats going on to make sure we dont regress here
![image](https://user-images.githubusercontent.com/5120858/166616662-3dd7f535-35e3-4c9c-bc5a-eba3183c1c46.png)

Reply:
From my testing setting CONFIG to Unsupported does not prevent the running of --cluster commands.
But setting the actual cluster command to Unsupported will naturally prevent the running of that --cluster command.

So I think there is no issue here.
If there are any --cluster commands that we can support then they will be supported.